# Praktische Übung 3: ML-Workflow

## Aufgabe 1

1. Laden Sie den Datensatz aus der Datei `titanic.csv` in einen Pandas-DataFrame. Die Datei befindet sich im `/data` Ordner auf GitHub.
2. Erstellen Sie einen neuen DataFrame, der nur die folgenden Spalten enthält: `Survived`, `Pclass`, `Age`, `Fare`, `Sibsp`, `Parch`.
3. Unterteilen Sie die Daten mit der Methode `train_test_split` in Trainings- und Testdaten (80% Training, 20% Test). Hinweis: Für die ersten drei Aufgaben können Sie den Code der Lösung der letzten Übung kopieren.
4. Überprüfen Sie welche Attribute `NaN`-Werte enthalten.
5. Füllen Sie diese Werte mit dem Durchschnittswert der übrigen Werte aus der gleichen Spalte auf.

1. Laden Sie den Datensatz aus der Datei `titanic.csv` in einen Pandas-DataFrame. Die Datei befindet sich im `/data` Ordner auf GitHub.

In [88]:
import pandas as pd
titanic = pd.read_csv('data/titanic.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

2. Erstellen Sie einen neuen DataFrame, der nur die folgenden Spalten enthält: `Survived`, `Pclass`, `Age`, `Fare`, `Sibsp`, `Parch`.

In [25]:
titanic_relevant = titanic[["Survived", "Pclass", "Age", "Fare", "SibSp", "Parch"]]
titanic_relevant.head()

Survived  Pclass   Age     Fare  SibSp  Parch
0         0       3  22.0   7.2500      1      0
1         1       1  38.0  71.2833      1      0
2         1       3  26.0   7.9250      0      0
3         1       1  35.0  53.1000      1      0
4         0       3  35.0   8.0500      0      0

3. Unterteilen Sie die Daten mit der Methode `train_test_split` in Trainings- und Testdaten (80% Training, 20% Test). Hinweis: Für die ersten drei Aufgaben können Sie den Code der Lösung der letzten Übung kopieren.

In [26]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    titanic_relevant.drop("Survived", axis=1),          # features
    titanic_relevant["Survived"],                       # label
    test_size=0.2,                                      # 20% test data
    random_state=42)                                    # fix random seed for reproducibility               

4. Überprüfen Sie welche Attribute `NaN`-Werte enthalten.

In [27]:
titanic_relevant.columns[titanic_relevant.isna().any()].tolist()    # check for NaN values

['Age']

5. Füllen Sie diese Werte mit dem Durchschnittswert der übrigen Werte aus der gleichen Spalte auf.

In [28]:
X_train["Age"].fillna(X_train["Age"].mean(), inplace=True)        # fill NaN values with mean

## Aufgabe 2

1. Skalieren Sie die Daten aus `x_train` mit einem `StandardScaler`.
2. Trainieren Sie eine logistische Regression auf den skalierten Daten und `y_train`.
3. Entfernen Sie alle Datenpunkte mit `NaN`-Werten aus `x_test`.
4. Entfernen Sie die genau gleichen Zeilen auch aus `y_test`.
5. Skalieren Sie die Datenpunkte `x_test` mit dem gleichen Skalierer wie in Aufgabe 2.
6. Machen Sie eine Vorhersage aus den skalierten Testdaten und berechnen Sie Accuracy, Precision und Recall.

1. Skalieren Sie die Daten aus `x_train` mit einem `StandardScaler`.

In [29]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)                           # fit scaler to training data
X_train_scaled = scaler.transform(X_train)                       # transform training data
X_train_scaled

array([[-1.61413602,  1.2322632 , -0.07868358, -0.47072241, -0.47934164],
       [-0.40055118, -0.50048197, -0.37714494, -0.47072241, -0.47934164],
       [ 0.81303367,  0.1926161 , -0.47486697, -0.47072241, -0.47934164],
       ...,
       [ 0.81303367,  0.88571416, -0.35580399,  1.23056874, -0.47934164],
       [-1.61413602, -1.19358003,  1.68320121,  0.37992316,  2.04874166],
       [-1.61413602, -0.65450376,  0.86074761, -0.47072241,  0.78470001]])

2. Trainieren Sie eine logistische Regression auf den skalierten Daten und `y_train`.

In [30]:
from sklearn.linear_model import LogisticRegression
reg = LogisticRegression()                         # create logistic regression object
reg.fit(X_train_scaled, y_train)                   # fit model to training data

LogisticRegression()

3. Entfernen Sie alle Datenpunkte mit `NaN`-Werten aus `x_test`.

In [31]:
X_test.dropna(inplace=True)                         # drop NaN values
X_test.shape                                        # check shape of test data

(142, 5)

4. Entfernen Sie die genau gleichen Zeilen auch aus `y_test`.

In [32]:
y_test = y_test[X_test.index]                        # drop all datapoints from test label that are not in test features
y_test.shape

(142,)

5. Skalieren Sie die Datenpunkte `x_test` mit dem gleichen Skalierer wie in Aufgabe 2.

In [33]:
X_test_scaled = scaler.transform(X_test)            # transform test data with scaler fitted to training data

6. Machen Sie eine Vorhersage aus den skalierten Testdaten und berechnen Sie Accuracy, Precision und Recall.

In [34]:
import sklearn.metrics as metrics
y_pred = reg.predict(X_test_scaled)                 # predict labels for test data
acc = metrics.accuracy_score(y_test, y_pred)        # calculate accuracy
prec = metrics.precision_score(y_test, y_pred)      # calculate precision
rec = metrics.recall_score(y_test, y_pred)          # calculate recall

print("Accuracy: ", acc)
print("Precision: ", prec)
print("Recall: ", rec)


Accuracy:  0.7464788732394366
Precision:  0.7894736842105263
Recall:  0.5172413793103449


## Aufgabe 3

1. Fügen Sie `Embarked` und `Sex` als weitere Feature hinzu.
2. Führen Sie ein one-hot-encoding dieser Attribute durch und durchlaufen Sie die Schritte aus Aufgabe 1 und 2, um das Modell mit den Features zu trainieren.
3. Vergleichen Sie die Performance des Modells mit und ohne diese Features.

1. Fügen Sie `Embarked` und `Sex` als weitere Feature hinzu.

In [36]:
import pandas as pd
titanic = pd.read_csv('data/titanic.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [37]:
titanic_relevant_2 = titanic[["Survived", "Pclass", "Sex", "Age", "Fare", "SibSp", "Parch", "Embarked"]]
titanic_relevant_2.head()

Survived  Pclass     Sex   Age     Fare  SibSp  Parch Embarked
0         0       3    male  22.0   7.2500      1      0        S
1         1       1  female  38.0  71.2833      1      0        C
2         1       3  female  26.0   7.9250      0      0        S
3         1       1  female  35.0  53.1000      1      0        S
4         0       3    male  35.0   8.0500      0      0        S

2. Führen Sie ein one-hot-encoding dieser Attribute durch und durchlaufen Sie die Schritte aus Aufgabe 1 und 2, um das Modell mit den Features zu trainieren.

In [40]:
Sex_One_Hot = pd.get_dummies(titanic_relevant_2["Sex"])                     # One-Hot-Encoding for Sex column
Embarked_One_Hot = pd.get_dummies(titanic_relevant_2["Embarked"])           # One-Hot-Encoding for Embarked column

titanic_relevant_One_Hot = pd.concat([titanic_relevant_2, Sex_One_Hot, Embarked_One_Hot], axis=1)   # concatenate One-Hot-Encoded columns to dataframe
titanic_relevant_One_Hot = titanic_relevant_One_Hot.drop(columns= ["Sex", "Embarked"])              # drop original columns
titanic_relevant_One_Hot.head()

Survived  Pclass   Age     Fare  SibSp  Parch  female  male  C  Q  S
0         0       3  22.0   7.2500      1      0       0     1  0  0  1
1         1       1  38.0  71.2833      1      0       1     0  1  0  0
2         1       3  26.0   7.9250      0      0       1     0  0  0  1
3         1       1  35.0  53.1000      1      0       1     0  0  0  1
4         0       3  35.0   8.0500      0      0       0     1  0  0  1

In [41]:
from sklearn.model_selection import train_test_split                     # split data into training and test data
X_train_2, X_test_2, y_train_2, y_test_2 = train_test_split(
    titanic_relevant_One_Hot.drop("Survived", axis=1),
    titanic_relevant_One_Hot["Survived"],
    test_size=0.2,
    random_state=42)

In [42]:
X_train_2["Age"].fillna(X_train_2["Age"].mean(), inplace=True)             # fill NaN values in the Age column with mean

In [43]:
from sklearn.preprocessing import StandardScaler
scaler_2 = StandardScaler().fit(X_train_2)                                  # fit scaler to training data
X_train_scaled_2 = scaler_2.transform(X_train_2)                            # transform training data
X_train_scaled_2

array([[-1.61413602,  1.2322632 , -0.07868358, ..., -0.46146201,
        -0.30335547,  0.59681695],
       [-0.40055118, -0.50048197, -0.37714494, ..., -0.46146201,
        -0.30335547,  0.59681695],
       [ 0.81303367,  0.1926161 , -0.47486697, ..., -0.46146201,
        -0.30335547,  0.59681695],
       ...,
       [ 0.81303367,  0.88571416, -0.35580399, ..., -0.46146201,
        -0.30335547,  0.59681695],
       [-1.61413602, -1.19358003,  1.68320121, ..., -0.46146201,
        -0.30335547,  0.59681695],
       [-1.61413602, -0.65450376,  0.86074761, ..., -0.46146201,
        -0.30335547,  0.59681695]])

In [44]:
from sklearn.linear_model import LogisticRegression
reg_2 = LogisticRegression()                            # create logistic regression object
reg_2.fit(X_train_scaled_2, y_train_2)                  # fit model to training data

LogisticRegression()

In [50]:
X_test_2.dropna(inplace=True)                # drop NaN values
print(X_test_2.shape)                        

y_test_2 = y_test_2[X_test_2.index]          # drop all datapoints from test label that are not in test features
print(y_test_2.shape)

(142, 10)
(142,)


In [51]:
X_test_scaled_2 = scaler_2.transform(X_test_2)      # transform test data with scaler fitted to training data

In [52]:
y_pred_2 = reg_2.predict(X_test_scaled_2)           # predict labels for test data

3. Vergleichen Sie die Performance des Modells mit und ohne diese Features.

In [54]:
import sklearn.metrics as metrics
acc_2 = metrics.accuracy_score(y_test_2, y_pred_2)
prec_2 = metrics.precision_score(y_test_2, y_pred_2)
rec_2 = metrics.recall_score(y_test_2, y_pred_2)

print("Acc w/o Sex and Embarked: ", acc , " Acc w/ Sex and Embarked: ", acc_2)
print("Prec w/o Sex and Embarked: ", prec, "Prec w/ Sex and Embarked: ", prec_2)
print("Recall w/o Sex and Embarked: ", rec, "Recall w/ Sex and Embarked: ", rec_2)

Acc w/o Sex and Embarked:  0.7464788732394366  Acc w/ Sex and Embarked:  0.8169014084507042
Prec w/o Sex and Embarked:  0.7894736842105263 Prec w/ Sex and Embarked:  0.7962962962962963
Recall w/o Sex and Embarked:  0.5172413793103449 Recall w/ Sex and Embarked:  0.7413793103448276


## Bonus

Entwerfen Sie ein Feature aus dem Attribut `Cabin` und schauen Sie, ob sich das Modell dadurch verbessert.

In [71]:
import pandas as pd
titanic = pd.read_csv('data/titanic.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [72]:
titanic_relevant_3 = titanic[["Survived", "Pclass", "Sex", "Age", "Fare", "SibSp", "Parch","Cabin", "Embarked"]]
titanic_relevant_3.head()

Survived  Pclass     Sex   Age     Fare  SibSp  Parch Cabin Embarked
0         0       3    male  22.0   7.2500      1      0   NaN        S
1         1       1  female  38.0  71.2833      1      0   C85        C
2         1       3  female  26.0   7.9250      0      0   NaN        S
3         1       1  female  35.0  53.1000      1      0  C123        S
4         0       3    male  35.0   8.0500      0      0   NaN        S

In [73]:
titanic_relevant_3["Cabin"].fillna("N", inplace=True)                               # fill NaN values with "N"
titanic_relevant_3["Cabin"] = titanic_relevant_3["Cabin"].apply(lambda x: x[0])     # keep only the first letter of the cabin
titanic_relevant_3.head()

C:\Users\s4nd3\AppData\Local\Temp\ipykernel_18604\524212744.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic_relevant_3["Cabin"].fillna("N", inplace=True)                               # fill NaN values with "N"
C:\Users\s4nd3\AppData\Local\Temp\ipykernel_18604\524212744.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic_relevant_3["Cabin"] = titanic_relevant_3["Cabin"].apply(lambda x: x[0])     # keep only the first letter of the cabin


Survived  Pclass     Sex   Age     Fare  SibSp  Parch Cabin Embarked
0         0       3    male  22.0   7.2500      1      0     N        S
1         1       1  female  38.0  71.2833      1      0     C        C
2         1       3  female  26.0   7.9250      0      0     N        S
3         1       1  female  35.0  53.1000      1      0     C        S
4         0       3    male  35.0   8.0500      0      0     N        S

In [87]:
# count the "N" values in the Cabin column
value_count = titanic_relevant_3["Cabin"].value_counts()
length = len(titanic_relevant_3["Cabin"])

# calculate the percentage of "N" values in the Cabin column
value_percentage = value_count/length

value_percentage

N    0.771044
C    0.066218
B    0.052750
D    0.037037
E    0.035915
A    0.016835
F    0.014590
G    0.004489
T    0.001122
Name: Cabin, dtype: float64

In [74]:
# One Hot Encoding for required columns
Sex_One_Hot_3 = pd.get_dummies(titanic_relevant_3["Sex"])
Embarked_One_Hot_3 = pd.get_dummies(titanic_relevant_3["Embarked"])
Cabin_One_Hot_3 = pd.get_dummies(titanic_relevant_3["Cabin"])

Embarked_One_Hot_3.rename(columns={"C": "Embarked_C", "Q": "Embarked_Q", "S": "Embarked_S"}, inplace=True) # rename columns so that there are no duplicates


titanic_relevant_One_Hot_3 = pd.concat([titanic_relevant_3, Sex_One_Hot_3, Embarked_One_Hot_3, Cabin_One_Hot_3], axis=1)    # concatenate One-Hot-Encoded columns to dataframe
titanic_relevant_One_Hot_3 = titanic_relevant_One_Hot_3.drop(columns= ["Sex", "Embarked", "Cabin"])                   # drop original columns
titanic_relevant_One_Hot_3.head()

Survived  Pclass   Age     Fare  SibSp  Parch  female  male  Embarked_C  \
0         0       3  22.0   7.2500      1      0       0     1           0   
1         1       1  38.0  71.2833      1      0       1     0           1   
2         1       3  26.0   7.9250      0      0       1     0           0   
3         1       1  35.0  53.1000      1      0       1     0           0   
4         0       3  35.0   8.0500      0      0       0     1           0   

   Embarked_Q  Embarked_S  A  B  C  D  E  F  G  N  T  
0           0           1  0  0  0  0  0  0  0  1  0  
1           0           0  0  0  1  0  0  0  0  0  0  
2           0           1  0  0  0  0  0  0  0  1  0  
3           0           1  0  0  1  0  0  0  0  0  0  
4           0           1  0  0  0  0  0  0  0  1  0

In [75]:
from sklearn.model_selection import train_test_split
X_train_3, X_test_3, y_train_3, y_test_3 = train_test_split(         # split data into training and test data
    titanic_relevant_One_Hot_3.drop("Survived", axis=1),
    titanic_relevant_One_Hot_3["Survived"],
    test_size=0.2,
    random_state=42)

In [76]:
X_train_3["Age"].fillna(X_train_3["Age"].mean(), inplace=True)      # fill NaN values in the Age column with mean

In [78]:
from sklearn.preprocessing import StandardScaler
scaler_3 = StandardScaler().fit(X_train_3)                # fit scaler to training data
X_train_scaled_3 = scaler_3.transform(X_train_3)          # transform training data
X_train_scaled_3

array([[-1.61413602,  1.2322632 , -0.07868358, ..., -0.06504853,
        -1.8649363 , -0.03750293],
       [-0.40055118, -0.50048197, -0.37714494, ..., -0.06504853,
         0.53621134, -0.03750293],
       [ 0.81303367,  0.1926161 , -0.47486697, ..., -0.06504853,
         0.53621134, -0.03750293],
       ...,
       [ 0.81303367,  0.88571416, -0.35580399, ..., -0.06504853,
         0.53621134, -0.03750293],
       [-1.61413602, -1.19358003,  1.68320121, ..., -0.06504853,
        -1.8649363 , -0.03750293],
       [-1.61413602, -0.65450376,  0.86074761, ..., -0.06504853,
        -1.8649363 , -0.03750293]])

In [79]:
from sklearn.linear_model import LogisticRegression
reg_3 = LogisticRegression()                  # create logistic regression object
reg_3.fit(X_train_scaled_3, y_train_3)        # fit model to training data

LogisticRegression()

In [80]:
X_test_3.dropna(inplace=True)          # drop NaN values from the test data
print(X_test_3.shape)

y_test_3 = y_test_3[X_test_3.index]     # drop all datapoints from test label that are not in test features
print(y_test_3.shape)

(142, 19)
(142,)


In [81]:
X_test_scaled_3 = scaler_3.transform(X_test_3)  	    # transform test data with scaler fitted to training data

In [82]:
y_pred_3 = reg_3.predict(X_test_scaled_3)               # predict labels for test data

In [83]:
import sklearn.metrics as metrics
acc_3 = metrics.accuracy_score(y_test_3, y_pred_3)
prec_3 = metrics.precision_score(y_test_3, y_pred_3)
rec_3 = metrics.recall_score(y_test_3, y_pred_3)

print("Acc w/o Cabin: ", acc_2 , " Acc w/ Cabin: ", acc_3)
print("Prec w/o Cabin: ", prec_2, "Prec w/ Cabin: ", prec_3)
print("Recall w/o Cabin: ", rec_2, "Recall w/ Cabin: ", rec_3)

Acc w/o Cabin:  0.8169014084507042  Acc w/ Cabin:  0.8098591549295775
Prec w/o Cabin:  0.7962962962962963 Prec w/ Cabin:  0.7818181818181819
Recall w/o Cabin:  0.7413793103448276 Recall w/ Cabin:  0.7413793103448276


-> no improvement by adding the Cabins, because in 77% of datapoints the Cabin value was not given